Update the "Application dependency management" recommendations

pradyunsg opened the pull request:
This is a more detailed recommendation that breaks up the recommendation
into multiple parts, to reflect that multiple tools need to be used in
workflows today to maximise compatibility with existing platforms.

It also recommends that deployed applications be isolated from the
system packages, as was implied but not explicitly noted.

The move away from Pipenv as the primary recommendation is a reflection
of the change that application development workflows have a greater
diversity in mature tooling available for them.

Toward #912

merwok commented on source/guides/tool-recommendations.rst:
```diff
@@ -11,17 +11,18 @@ what tools are currently recommended, then here it is.
-
-* :ref:`pip`
+* Use :ref:`virtualenv`, or :doc:`venv <python:library/venv>` to isolate application
+  specific dependencies from a shared Python installation. [4]_
```
Now that virtualenv uses venv when possible, maybe mentioning only virtualenv here would be a clearer guideline?

pradyunsg replied:
This mirrors the phrasing in the next section of recommendations. If we want to change this, I'd prefer we do it in a follow up that updates both.

bhrutledge commented on source/guides/tool-recommendations.rst:
```diff
@@ -11,17 +11,18 @@ what tools are currently recommended, then here it is.
+  dependencies during deployment.
 
-* :ref:`pip`
+* Use :ref:`virtualenv` or :doc:`venv <python:library/venv>` to isolate
+  application-specific dependencies from a shared Python installation. [4]_
```
This item is duplicated in "Installation tool recommendations". Is it necessary here?

pradyunsg replied:
Yes. From the commit message:

> It also recommends that deployed applications be isolated from the
system packages, as was implied but not explicitly noted.

bhrutledge reviewed (commented):
Thanks for this. I took the liberty of pushing some copy edits, and I have one content question.

pradyunsg commented on source/guides/tool-recommendations.rst:
```diff
@@ -11,14 +11,14 @@ what tools are currently recommended, then here it is.
 
-* Use :ref:`pip` in a `secure manner`_ to install your distribution
-  packages, when deploying Python applications.
+* Use :ref:`pip` in a `secure manner`_ to install a Python application's
```
This changed meaning, since this now excludes the application itself instead of including it in this description.

bhrutledge commented on source/guides/tool-recommendations.rst:
```diff
@@ -11,17 +11,18 @@ what tools are currently recommended, then here it is.
+  application-specific dependencies.
 
-When ``pipenv`` does not meet your use case, consider other tools like:
+* Use :ref:`pip` in a `secure manner`_ to install a Python application and its
+  dependencies during deployment.
```
@pradyunsg How's this?

pradyunsg replied:
Works! Thanks!

pradyunsg commented:
@bhrutledge I've rebased this on `main` and squashed your copy edits to the new content (dropping the reordering). You can click on "force-pushed" text in the GitHub event above, to review what exactly has changed across the force-push.
